cli/src/org/partiql/shell/ShellHighlighter.kt

```diff
@@ -11,32 +11,134 @@
+            val lineNumber =
+                e.errorContext[Property.LINE_NUMBER]?.longValue()?.toInt() ?: return builder.toAttributedString()
+            val token = tokens.find {
+                it.span.column.toInt() == column && it.span.line.toInt() == lineNumber
+            } ?: return builder.toAttributedString()
+            builder = createAttributeStringBuilder(token, lineIndexesMap, builder, true)
+        }
+        return builder.toAttributedString()
+    }
 
     override fun setErrorPattern(errorPattern: Pattern?) {}
 
     override fun setErrorIndex(errorIndex: Int) {}
+
+    /**
+     * Based on the [token] type and location, this function will return a replica of the [stringBuilder] with the
+     * token's location having an updated color-scheme (style). If [isFailure] is passed, the style will always be RED.
+     */
+    private fun createAttributeStringBuilder(
+        token: Token,
+        lineIndexes: Map<Int, Int>,
+        stringBuilder: AttributedStringBuilder,
+        isFailure: Boolean
+    ): AttributedStringBuilder {
+        val a = AttributedStringBuilder()
+        val style = when (isFailure) {
+            true -> AttributedStyle().foreground(AttributedStyle.RED)
+            false -> getStyle(token)
+        }
+        val column = token.span.column.toInt()
+        val lineNumber = token.span.line.toInt() - 1
+        val length = token.span.length.toInt()
+        val index = lineIndexes[lineNumber]?.plus(column) ?: return stringBuilder
+        a.append(stringBuilder.subSequence(0, index))
+        a.append(stringBuilder.subSequence(index, index + length), style)
+        a.append(stringBuilder.subSequence(index + length, stringBuilder.length))
+        return a
+    }
+
+    /**
+     * Sets the color and thickness of the string based on the [token] type
+     */
+    private fun getStyle(token: Token): AttributedStyle {
+        var style = AttributedStyle()
+        val attrCode = when (token.type) {
+            TokenType.KEYWORD -> if (token.isDataType) AttributedStyle.GREEN else AttributedStyle.CYAN
+            TokenType.AS, TokenType.FOR, TokenType.ASC, TokenType.LAST, TokenType.DESC,
+            TokenType.BY, TokenType.FIRST -> AttributedStyle.CYAN
+            TokenType.LITERAL -> if (token.value is IonString) AttributedStyle.YELLOW else AttributedStyle.BLUE
+            TokenType.ION_LITERAL -> AttributedStyle.YELLOW
+            TokenType.OPERATOR -> AttributedStyle.WHITE
+            TokenType.QUOTED_IDENTIFIER -> AttributedStyle.BRIGHT
+            TokenType.IDENTIFIER -> AttributedStyle.BRIGHT
+            TokenType.MISSING -> AttributedStyle.BLUE
+            TokenType.NULL -> AttributedStyle.BLUE
+            else -> AttributedStyle.WHITE
+        }
+        style = style.foreground(attrCode)
+
+        return when (token.type) {
+            TokenType.IDENTIFIER, TokenType.OPERATOR -> style.bold()
+            else -> style
+        }
+    }
```

Comment:
Consider putting these in a map. Kotlin `when` is pretty clean, but it just leads to a lot of branches for what is just a lookup

Reply:
See my comment about syntax files. If okay, I'll make this a requirement for the next issue to define a syntax file for users to pass in. Then, we can move this to a standardized file.